music_assistant/providers/nicovideo/helpers/utils.py

```diff
@@ -0,0 +1,93 @@
+"""Utility functions for handling cookies and converting them into Netscape format."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from mashumaro import DataClassDictMixin
+
+# Playlist, Album, and Track cannot be placed under TYPE_CHECKING
+# because they are used at runtime by DataClassDictMixin
+from music_assistant_models.media_items import (
+    Album,
+    AudioFormat,
+    Playlist,
+    Track,
+)
+
+from music_assistant.constants import VERBOSE_LOG_LEVEL
+from music_assistant.providers.nicovideo.constants import (
+    NICOVIDEO_AUDIO_BIT_DEPTH,
+    NICOVIDEO_AUDIO_CHANNELS,
+    NICOVIDEO_CODEC_TYPE,
+    NICOVIDEO_CONTENT_TYPE,
+)
+
+if TYPE_CHECKING:
+    import logging
+
+    from music_assistant.models.music_provider import MusicProvider
+
+
+@dataclass
+class PlaylistWithTracks(DataClassDictMixin):
+    """Helper class to hold playlist and its tracks."""
+
+    playlist: Playlist
+    tracks: list[Track]
+
+
+@dataclass
+class AlbumWithTracks(DataClassDictMixin):
+    """Helper class to hold album and its tracks."""
+
+    album: Album
+    tracks: list[Track]
+
+
+async def cache_track(provider: MusicProvider, track: Track) -> None:
+    """Cache single track with provider item cache.
+
+    Note: While MusicAssistant's get_provider_item automatically handles cache retrieval
+    and storage, this helper function is needed for explicit cache updates (e.g., when
+    adding album information to tracks) since MusicAssistant doesn't provide a dedicated
+    cache-only update function.
+    """
+    cache_key = f"track.{track.item_id}"
+    await provider.mass.cache.set(
+        cache_key,
+        track.to_dict(),
+        provider=provider.lookup_key,
+    )
```

Comment:
Note that we recently did a bunch of changes to caching, basically we removed "automagic" caching by default from the core controllers making it a responsibility of the providers - which know better what and how long to cache.

Also we added support for a cache category to more fine grained split up the results in the cache (and it will give a bit of s speed boost due the indexes)